src/main/scala/is/hail/keytable/KeyTable.scala

```diff
@@ -377,29 +311,28 @@ case class KeyTable(@transient hc: HailContext, @transient rdd: RDD[(Annotation,
     val localNFields = nFields
     val localTypes = fields.map(_.typ)
 
-    KeyTable.toSingleRDD(rdd, nKeys, nValues)
-      .mapPartitions { it =>
-        val sb = new StringBuilder()
-        val nulls: Seq[Any] = Array.fill[Annotation](localNFields)(null)
+    rdd.mapPartitions { it =>
+      val sb = new StringBuilder()
+      val nulls: Seq[Any] = Array.fill[Annotation](localNFields)(null)
 
-        it.map { r =>
-          sb.clear()
+      it.map { r =>
+        sb.clear()
 
-          val r2 =
-            if (r == null)
-              nulls
-            else
-              r.asInstanceOf[Row].toSeq
-          assert(r2.length == localTypes.length)
+        val r2 =
+          if (r == null)
+            nulls
+          else
+            r.asInstanceOf[Row].toSeq
+        assert(r2.length == localTypes.length)
 
-          r2.zip(localTypes)
-            .foreachBetween { case (x, t) =>
-              sb.append(t.str(x))
-            }(sb += '\t')
+        r2.zip(localTypes)
```

Comment:
This copies the data.  Use iterators, or better, use `(0 until localTypes.length).foreachBetween { . . . } `